flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/InternalTimerServiceImpl.java

```diff
@@ -296,7 +302,9 @@ public void advanceWatermark(long time) throws Exception {
 
         InternalTimer<K, N> timer;
 
-        while ((timer = eventTimeTimersQueue.peek()) != null && timer.getTimestamp() <= time) {
+        while ((timer = eventTimeTimersQueue.peek()) != null
+                && timer.getTimestamp() <= time
+                && !cancellationContext.isCancelled()) {
```

Comment:
I would add the same check 10 lines above in the `onProcessingTimer()` method (and a test for it)